samples: net: dns_resolve: Improve DHCPv4 handling

rlubos opened the pull request:
There is some ambiguity with the DHCPv4 handling in the dns_resolve sample. On one hand, the sample uses net_config library, which does initiate the DHCPv4 on the interface (if enabled) and may block the initialization until the address is assigned (in case there is no other statically assigned IPv4 addresses.) On the other hand, the sample registers for NET_EVENT_IPV4_ADDR_ADD in case DHCPv4 is in use, delaying the DNS queries until address is assigned.

In case net_config delayed the initialization however, this will not work, as the event handler will be registered only after DHCPv4 address is assigned, so the callback will not get executed.

Fix this, by checking if DHCPv4 assigned address already exists on an interface - if so, schedule DNS queries immediately. Otherwise (for example in case when net_config was not configured to wait for address) register an event callback, and schedule queries from there.

rlubos committed:
samples: net: dns_resolve: Improve DHCPv4 handling

There is some ambiguity with the DHCPv4 handling in the dns_resolve
sample. On one hand, the sample uses net_config library, which does
initiate the DHCPv4 on the interface (if enabled) and may block the
initialization until the address is assigned (in case there is no other
statically assigned IPv4 addresses.) On the other hand, the sample
registers for NET_EVENT_IPV4_ADDR_ADD in case DHCPv4 is in use, delaying
the DNS queries until address is assigned.

In case net_config delayed the initialization however, this will not
work, as the event handler will be registered only after DHCPv4 address
is assigned, so the callback will not get executed.

Fix this, by checking if DHCPv4 assigned address already exists on an
interface - if so, schedule DNS queries immediately. Otherwise (for
example in case when net_config was not configured to wait for address)
register an event callback, and schedule queries from there.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on samples/net/dns_resolve/src/main.c:
```diff
@@ -192,17 +203,24 @@ static void ipv4_addr_add_handler(struct net_mgmt_event_callback *cb,
 
 static void setup_dhcpv4(struct net_if *iface)
 {
+	for (int i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+		struct net_if_addr *if_addr =
+			&iface->config.ip.ipv4->unicast[i];
+
+		if (if_addr->addr_type != NET_ADDR_DHCP || !if_addr->is_used) {
+			continue;
+		}
+
+		/* Already have DHCP assigned address, schedule queries. */
+		schedule_ipv4_queries();
+		return;
+	}
+
```
I am just wondering that should we have an API function (similar to `net_if_foreach()` so that user supplied callback could be called for each address) that could provide this info. Now we are accessing net_if internals from sample which we should try to avoid.

rlubos replied:
That could be useful, I'll try to introduce such function instead.

rlubos replied:
@jukkar I've added helper functions to loop over IP addresses configured on an interface and made use of it in the sample.

rlubos committed:
net: if: Add functions to loop over IPv4/IPv6 addresses

Add new net_if API functions which allow to loop over all valid
IPv4/IPv6 addresses assigned to the interface and execute a callback
function on them.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: if: Add tests to verify net_if_ipv4/6_addr_foreach()

Add tests cases which verify that net_if_ipv4/6_addr_foreach() executes
callbacks properly on assigned addresses.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (dismissed):
Thanks, LGTM.

rlubos commented:
Added additional comments in the code as per @glarsennordic suggestion.
